File: infra/components/workflows.py
"""GCP Workflows for genomics pipeline orchestration.

Deploys three workflow definitions (biomarker discovery, sample QC,
prompt optimization) with a dedicated service account and IAM bindings.
"""

from pathlib import Path

import pulumi
import pulumi_gcp as gcp

# Resolve workflow YAML definitions relative to project root
_PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent
_WORKFLOW_DEFS = _PROJECT_ROOT / "workflows" / "definitions"


class GenomicsWorkflows(pulumi.ComponentResource):
    """GCP Workflows with service account, IAM, and three workflow definitions."""

    biomarker_discovery_id: pulumi.Output[str]
    sample_qc_id: pulumi.Output[str]
    prompt_optimization_id: pulumi.Output[str]

    def __init__(
        self,
        name: str,
        project_id: str,
        region: str,
        activity_service_url: pulumi.Input[str],
        opts: pulumi.ResourceOptions | None = None,
    ):
        super().__init__("genomics:infra:GenomicsWorkflows", name, None, opts)

        child = pulumi.ResourceOptions(parent=self)

        # Service account for workflow execution
        sa = gcp.serviceaccount.Account(
            f"{name}-sa",
            project=project_id,
            account_id="precision-genomics-workflows",
            display_name="Precision Genomics Workflows",
            opts=child,
        )

        # IAM: allow the SA to invoke workflows and Cloud Run
        for role_suffix, role in [
            ("workflows-invoker", "roles/workflows.invoker"),
            ("run-invoker", "roles/run.invoker"),
        ]:
            gcp.projects.IAMMember(
                f"{name}-{role_suffix}",
                project=project_id,
                role=role,
                member=sa.email.apply(lambda e: f"serviceAccount:{e}"),
                opts=child,
            )

        # Deploy three workflow definitions
        workflow_specs = {
            "biomarker-discovery": "biomarker_discovery.yaml",
            "sample-qc": "sample_qc.yaml",
            "prompt-optimization": "prompt_optimization.yaml",
        }

        workflows: dict[str, gcp.workflows.Workflow] = {}
        for wf_name, yaml_file in workflow_specs.items():
            source = (_WORKFLOW_DEFS / yaml_file).read_text()
            workflows[wf_name] = gcp.workflows.Workflow(
                f"{name}-{wf_name}",
                name=f"precision-genomics-{wf_name}",
                project=project_id,
                region=region,
                service_account=sa.id,
                source_contents=source,
                opts=child,
            )

        self.biomarker_discovery_id = workflows["biomarker-discovery"].id
        self.sample_qc_id = workflows["sample-qc"].id
        self.prompt_optimization_id = workflows["prompt-optimization"].id

        self.register_outputs(
            {
                "biomarker_discovery_id": self.biomarker_discovery_id,
                "sample_qc_id": self.sample_qc_id,
                "prompt_optimization_id": self.prompt_optimization_id,
            }
        )

File: infra-ts/components/workflows.ts
import * as pulumi from "@pulumi/pulumi";
import * as gcp from "@pulumi/gcp";
import * as fs from "fs";
import * as path from "path";

export class GenomicsWorkflows extends pulumi.ComponentResource {
  public readonly biomarkerDiscoveryId: pulumi.Output<string>;
  public readonly sampleQcId: pulumi.Output<string>;
  public readonly promptOptimizationId: pulumi.Output<string>;

  constructor(
    name: string,
    args: {
      projectId: string;
      region: string;
      activityServiceUrl: pulumi.Input<string>;
    },
    opts?: pulumi.ComponentResourceOptions,
  ) {
    super("genomics:infra:GenomicsWorkflows", name, {}, opts);
    const child = { parent: this };

    // Service account for workflows
    const sa = new gcp.serviceaccount.Account(`${name}-sa`, {
      accountId: "genomics-workflows",
      project: args.projectId,
      displayName: "Genomics Workflows Service Account",
    }, child);

    // Grant Cloud Run invoker role
    new gcp.projects.IAMMember(`${name}-sa-run-invoker`, {
      project: args.projectId,
      role: "roles/run.invoker",
      member: pulumi.interpolate`serviceAccount:${sa.email}`,
    }, child);

    // Load workflow YAML files
    const workflowsDir = path.join(__dirname, "..", "..", "workflows", "definitions");
    const loadYaml = (filename: string): string => {
      try {
        return fs.readFileSync(path.join(workflowsDir, filename), "utf-8");
      } catch {
        return `main:\n  steps:\n    - init:\n        assign:\n          - status: "placeholder"`;
      }
    };

    const biomarkerWorkflow = new gcp.workflows.Workflow(`${name}-biomarker`, {
      name: "biomarker-discovery",
      project: args.projectId,
      region: args.region,
      serviceAccount: sa.email,
      sourceContents: loadYaml("biomarker_discovery.yaml"),
    }, child);

    const sampleQcWorkflow = new gcp.workflows.Workflow(`${name}-sample-qc`, {
      name: "sample-qc",
      project: args.projectId,
      region: args.region,
      serviceAccount: sa.email,
      sourceContents: loadYaml("sample_qc.yaml"),
    }, child);

    const promptOptWorkflow = new gcp.workflows.Workflow(`${name}-prompt-opt`, {
      name: "prompt-optimization",
      project: args.projectId,
      region: args.region,
      serviceAccount: sa.email,
      sourceContents: loadYaml("prompt_optimization.yaml"),
    }, child);

    this.biomarkerDiscoveryId = biomarkerWorkflow.name;
    this.sampleQcId = sampleQcWorkflow.name;
    this.promptOptimizationId = promptOptWorkflow.name;

    this.registerOutputs({
      biomarkerDiscoveryId: this.biomarkerDiscoveryId,
      sampleQcId: this.sampleQcId,
      promptOptimizationId: this.promptOptimizationId,
    });
  }
}
